### PR TITLE
#927: add export_session callback on save, set initial project version=1

### PR DESCRIPTION
**Pull request by aparna-d2**

### CODE REVIEW

File: engine.py
Comment (aparna-d2) [inline engine.py]: There is no function in mari to check what callbacks are already connected to a signal...but disconnecting just returns False if the connection doesn't exist. Seems safe to do this.
__on_project_opened() just tries to switch context. Execution of this multiple times is avoided by checking if current and changed context are the same. But we cannot do that for file save (we want to overwrite even if file exists, because this is a WIP save)

File: engine.py
Comment (aparna-d2) [inline engine.py]: For backwards compatibility with already created projects that have no version set.

Comment (mbrainerd) [reply engine.py]: Do we really want to export a session on every File Save? This is a different paradigm than every other engine, which typically keeps the native File Save functionality and would do the `exportSession()` inside of the tk-multi-workfiles save callback in scene_operation_{engine_name}.py. Or is this in-lieu of workfiles?

File: python/tk_mari/project.py
Comment (mbrainerd) [inline python/tk_mari/project.py]: Why not just call `set_project_version()` instead?

**Review by mbrainerd — CHANGES_REQUESTED** [review]
See inline comments

File: python/tk_mari/project.py
Comment (aparna-d2) [reply python/tk_mari/project.py]: Oh, yes, we can do that too. Will update.

File: engine.py
Comment (aparna-d2) [reply engine.py]: The reason I am doing this is to make the exported session (msf file) act like a work file.
The ’File Save’ in tk-multi-workfiles would version up, whereas this is meant to save the current version.
Typically, say in maya, you save v001.ma and every time you do a native file save, it would overwrite v001 with the current session data. That is what I am trying to achieve here (save the same version, not update)
This would mean I can base the rest of the operations around msf files and work the same way as other DCCs (file open -> import session (create project + import geo if required), file save -> version up + export session, publish -> copy current workfile)

File: engine.py
Comment (aparna-d2) [inline engine.py:135]: I believe the metadata setting should be handled by projectmanager (or workfiles, if that is enabled)
So exposed a public method to set_metadata and removed this from post_app_init and post_context_change.

File: engine.py
Comment (mbrainerd) [inline engine.py:464]: Do not set a default version. If we fail work template validation then we should error out.

**Review by mbrainerd — APPROVED** [review]
Approved pending minor change request.

File: engine.py
Comment (aparna-d2) [reply engine.py:464]: The thing is, since v3 changes are already in use, and I am only setting new project version as 1 starting this pull request, earlier projects may have other metadata but not version (which should ideally have been 1) That's why I mentioned this is for backwards compatibility (in Github comment, not code)

Comment (mbrainerd) [reply engine.py:464]: Then I would suggest actually setting the metadata in the `__on_project_opened` callback assuming the project is SGTK-ified already and just missing its version metadata.

Comment (aparna-d2) [reply engine.py:464]: That works too.

File: engine.py
Comment (aparna-d2) [inline engine.py]: projectmanager doesn't need to call this as `engine.create_project()` already does that. Should I leave this exposed or remove it?
Ideally, metadata should be set on project creation. I can't think of a reason we would want to switch metadata later (updating version is already handled. Otherwise, a project should stick with a single context...?)

Comment (mbrainerd) [reply engine.py]: Yes, I can't think of a reason to keep this exposed either.

**Comment by aparna-d2** [comment]
Looks like I don't have write access... @mbrainerd could you merge/grant me permission?

File: engine.py
Comment (aparna-d2) [reply engine.py]: Removed.

**Review by mbrainerd — APPROVED** [review]

**Pull request merged**
